text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,1276 @@
+- Encourage progress and avoid stagnation
+  - Proactively upgrading means the total benefit to developers from investments made in Rust is higher
+  - Conversely, when most of the community is on old versions, it has a chilling effect on improving Rust
+  - This also means feedback can come more quickly, making it easier and cheaper to pivot with user needs
+  - Spreading the cost of upgrades over time makes forced-upgrades (e.g. for a security vulnerability) less of an emergency
+  - Our commitment to compatibility helps keep the cost of upgrade low
+- When not competing with the above, we should do the right thing for the user rather than disrupt their flow to telling them what they should instead do
+
+And keeping in mind
+- The Rust project only supports the latest version
+  (e.g bug and security fixes)
+  and the burden for support for older versions is on the vendor providing the older Rust toolchain.
+- Even keeping upgrade costs low, there is still a re-validation cost that mission critical applications must pay
+- Dependencies in `Cargo.lock` are not expected to change from contributors using different versions of the Rust toolchain without an explicit action like changing `Cargo.toml` or running `cargo update`
+  - e.g. If the maintainer does `cargo add foo && git commit && git push`,
+    then a contributor doing `git pull && cargo check` should not have a different selection of dependencies, independent of their toolchain versions (which might mean the second user sees an error about an incompatible package).
+
+Some implications:
+- "Support" in MSRV implies the same quality and responsiveness to bug reports, regardless of Rust version
+- MSRV applies to all interactions with a project
+  (including registry dependency, git dependency, `cargo install`, contributor experience),
+  unless documented otherwise like
+  - Some projects may document that enabling a feature will affect the MSRV (e.g. [moka](https://docs.rs/moka/0.12.1/moka/#minimum-supported-rust-versions))
+  - Some projects may have a higher MSRV for building the repo (e.g. `Cargo.lock` with newer dependencies, reliance on cargo features that get stripped on publish)
+- We should focus the cost for maintaining support for older versions of Rust on the user of the old version and away from the maintainer or the other users of the library or tool
+  - Costs include lower developer productivity due to lack of access to features,
+    APIs that don't integrate with the latest features,
+    and slower build times due to pulling in extra code to make up for missing features
+    (e.g. clap dropping its dependency on
+    [is-terminal](https://crates.io/crates/is-terminal) in favor of
+    [`IsTerminal`](https://doc.rust-lang.org/std/io/trait.IsTerminal.html)
+    cut build time from [6s to 3s](https://github.com/rosetta-rs/argparse-rosetta-rs/commit/378cd2c30679afdf9b9843dbadea3e8951090809))
+
+### Latest Rust with no MSRV
```

Comment:
@djc (moved to make it easier to follow the thread)

> I definitely have some disagreement with the underlying principles, but one problem with the suggestions laid out in the initial post is that the RFC requires ploughing through a lot of text (rather than an attempt at concisely laying out the underlying design) to explain these principles themselves. I think there are roughly 4 quadrants of stakeholders:
> 
>     * Users:
>       
>       1. Users who stick to stable or a recently stable version
>       2. Users who stick to a more conservative MSRV
> 
>     * Maintainers:
>       3. Maintainers who don't want to care about the supported Rust version/have a liberal MSRV policy
>       4. Maintainers who want to support a relatively conservative MSRV
> 
> 
> The principles in the RFC don't seem to make as much of a distinction between users and maintainers, whereas I think this is an important distinction, as the groups have limited overlap, at least in the sense that most users are not maintainers. And, while there are many more users, we should highly value the maintainers because of the value they bring to the Rust ecosystem.
> 
> I think the changes here are intended mainly to improve the lives for those in group 2 -- users who are, for whatever reason, on an older version of Rust. To me it is important that the experience for group 4 (which I consider myself to be a part of) is not regressed, because I think a number of experienced maintainers of popular crates are in this category. Instead, you've prioritized this group (as far as I can tell) at priority 4, below all the other groups, which doesn't seem great.
> 
> In earlier discussions, I posited that this might be because there is a difference between maintaining low-level libraries (which have a larger audience, and thus more constraints on evolving compatibility) and higher-level libraries or applications. (I also think this matches with the actual people pushing back on this FCP.)

Reply:
Maintainers can find themselves in either "Extended MSRV" or "Extended published MSRV w/ latest development MSRV".  For example, I do the former.

The main case I've seen for made for "Extended published MSRV w/ latest development MSRV" is that it enables maintainers to use the latest dependencies.  However, the use of the latest dependencies is restricted to behavior of the dependencies and not their API.  For me, the discussion of these benefits has been hand-wavy and it would help a lot for people to call out how they have been benefited and to describe why that anecdote is generally applicable enough for what the default is for the Rust community.

I also find myself jumping ahead and because there are points being made about costs related to that workflow, like being able to validate patches you are making to your dependencies.  This gets into the question of what "support" in MSRV means.

Reply:
>  For me, the discussion of these benefits has been hand-wavy and it would help a lot for people to call out how they have been benefited and to describe why that anecdote is generally applicable enough for what the default is for the Rust community.

To clarify, are you asking for an explanation of why I (and probably others) develop for an MSRV without using the toolchain associated with the MSRV in my development environment?

For me, I work on many different projects with different MSRVs but I've found the tooling good enough that I don't need to switch toolchains in my local development environment. I don't use `rust-toolchain.toml` files and mostly rely on CI jobs to check MSRV for me, and this seems like the path that requires the least effort to maintain to me.

Reply:
The question is not "why do you use the latest toolchain?".  The RFC actually encourages people to use the latest toolchain and some downsides for alternatives is that I believe it will push people to specifying a `rust-toolchain.toml` file.

I've seen a lot of value add in using the latest toolchain.  What I do not see is enough value in using the latest, MSRV-incompatible dependencies for day to day development.  

If I'm developing a library, I feel like most concerns about being "behind" seem negligible because I'm not a product on its own.

If I'm developing an application, then I should be using the same dependencies for testing and distribution.
- If I'm distributing through `cargo install`, then `--locked` is expected to contain your MSRV dependencies
- If I'm distributing using a verified toolchain which is why I have an MSRV, then I should be developing with the same dependencies as I distribute
- That just leaves when I distribute using the latest toolchain but I want to allow others to build from source with older toolchains.  If people are in this state, I'd want to better understand their situation (and not just hypotheticals we can come up with about it)